File: Code Generation/responses/cpp/deepseek/hard/count-beautiful-numbers.cpp
#include <vector>
#include <string>
#include <functional>
#include <cstring>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long beautifulNumbers(int l, int r) {
        return count_le(r) - count_le(l - 1);
    }

    long long count_le(long long n) {
        if (n == 0) return 0;
        string s = to_string(n);
        int len = s.length();
        
        // product range: min product = 1 (digit 1), max product = 9^len
        int max_product = 1;
        for (int i = 0; i < len; i++) max_product *= 9;
        
        // sum range: min sum = 1 (digit 1), max sum = 9*len
        int max_sum = 9 * len;
        
        // DP: pos, sum, product, tight, started
        // product can be large, but we only care about divisibility
        // Use memoization with product reduced by gcd(product, sum) * product / sum?
        // Actually product % sum == 0 is equivalent to product/sum is integer
        // But product can be up to 9^9 ~ 387M, too large for DP array
        // Instead track product and sum, and check divisibility at end
        
        // We'll use direct DP with map memoization for product
        // But we can optimize: product % sum == 0 only depends on product mod sum
        // Actually we need product mod sum == 0, which is product % sum == 0
        // So we need to track product % sum, but sum changes, so need both
        
        // Use memoization with map
        // key: (pos, sum, product_mod_sum? no, product itself is large)
        // We'll use direct product with limits
        
        // Since digits <= 9, product grows exponentially, but state space manageable
        // Let's use DP with memo[10][9*10+1][product?] but product huge
        // Instead use vector of maps for product values
        // Or do digit DP that checks at leaf nodes
        
        vector<vector<unordered_map<int, long long>>> memo(len + 1, 
            vector<unordered_map<int, long long>>(max_sum + 1));
        
        function<long long(int, int, int, bool, bool)> dfs = [&](int pos, int sum, int prod, bool tight, bool started) -> long long {
            if (pos == len) {
                if (started && sum > 0 && prod % sum == 0) return 1;
                return 0;
            }
            if (!tight && started && memo[pos][sum].count(prod)) {
                return memo[pos][sum][prod];
            }
            int limit = tight ? s[pos] - '0' : 9;
            long long res = 0;
            for (int dig = 0; dig <= limit; dig++) {
                bool new_tight = tight && (dig == limit);
                if (!started && dig == 0) {
                    // not started yet
                    res += dfs(pos + 1, 0, 1, new_tight, false);
                } else {
                    int new_sum = sum + dig;
                    int new_prod = (started ? prod * dig : dig);
                    if (new_prod == 0) new_prod = 1; // 0 means product is 0
                    res += dfs(pos + 1, new_sum, new_prod, new_tight, true);
                }
            }
            if (!tight && started) {
                memo[pos][sum][prod] = res;
            }
            return res;
        };
        
        return dfs(0, 0, 1, true, false);
    }
};

File: Code Generation/responses/python/deepseek/hard/count-beautiful-numbers.py
class Solution:
    def beautifulNumbers(self, l: int, r: int) -> int:
        from functools import lru_cache
        
        def count_up_to(n):
            if n == 0:
                return 0
            digits = list(map(int, str(n)))
            m = len(digits)
            
            # The maximum sum of digits for numbers up to 1e9 is 9*9 = 81
            # The maximum product of digits is also 9^9 = 387420489, but we cannot store all.
            # We'll store product and sum in the DP state.
            # Since product can be large, we'll use the fact that we only need to know
            # whether product % sum == 0. We can store product % sum in the state,
            # but sum changes dynamically. So we need to iterate over possible sums.
            
            # Precompute all possible sums (from 1 to 81)
            @lru_cache(None)
            def dp(pos, tight, sum_digits, prod_digits):
                if pos == m:
                    if sum_digits > 0:
                        return 1 if prod_digits % sum_digits == 0 else 0
                    return 0
                
                limit = digits[pos] if tight else 9
                total = 0
                for d in range(limit + 1):
                    if d == 0 and pos == 0:
                        # leading zero skip product/sum
                        new_sum = sum_digits
                        new_prod = prod_digits if prod_digits != 0 else 1
                    else:
                        new_sum = sum_digits + d
                        if prod_digits == 0:
                            new_prod = d
                        else:
                            new_prod = prod_digits * d
                    
                    total += dp(pos + 1, tight and d == limit, new_sum, new_prod)
                return total
            
            # But this DP will be too slow because product can be large.
            # We need a better approach: iterate over all possible sums of digits S (1 to 81)
            # and count numbers where sum of digits = S and product % S == 0.
            
            # We'll do digit DP with state (pos, tight, sum_mod, product_mod, sum_target)
            # But we also need to know if we are tracking sum to a target.
            # Alternative: iterate over target sum from 1 to 81 (max sum for 9 digits)
            # For each target sum, we do DP counting numbers where:
            #   sum_of_digits = target_sum, and product_of_digits % target_sum == 0
            # Then sum over all target sums.
            
            max_sum = 9 * m
            total_count = 0
            
            for target_sum in range(1, max_sum + 1):
                @lru_cache(None)
                def dp_target(pos, tight, current_sum, current_prod_mod):
                    if pos == m:
                        return 1 if (current_sum == target_sum and current_prod_mod == 0) else 0
                    
                    limit = digits[pos] if tight else 9
                    total = 0
                    for d in range(limit + 1):
                        new_tight = tight and d == limit
                        new_sum = current_sum + d
                        if new_sum > target_sum:
                            continue
                        if d == 0 and pos == 0:
                            new_prod_mod = 0 if current_prod_mod == 0 else 1
                        else:
                            if current_prod_mod == 0:
                                new_prod_mod = d % target_sum
                            else:
                                new_prod_mod = (current_prod_mod * d) % target_sum
                        total += dp_target(pos + 1, new_tight, new_sum, new_prod_mod)
                    return total
                
                total_count += dp_target(0, True, 0, 0)
                dp_target.cache_clear()
            
            return total_count
        
        return count_up_to(r) - count_up_to(l - 1)